### PR TITLE
Support coverage_ignore_py in the coverage extension

### PR DESCRIPTION
**Pull request by Gallaecio**
Subject: Extend the existing coverage extension with an option that allows ignoring Python API members regardless of their type, and based on their full import path.

### Feature or Bugfix
- Feature

### Purpose
Currently, the coverage extension only allows ignoring Python API members based on their base name and type. This means that you cannot ignore a specific attribute name on a specific class, it’s either all or none. It also mean that you cannot use a single expression to ignore Python API members of different types. You cannot ignore specific class attributes either.

This new setting allows doing all of it with a simple list of regular expressions that are checked against the full import path of any Python API member.

### Relates
- https://github.com/scrapy/scrapy/pull/3567

### CODE REVIEW

**Comment by Gallaecio** [comment]
I looked into the existing tests, to extend them, but they did not seem to cover the existing type-specific options, and the comments seem to suggest that testing this extensions is rather difficult. Please, let me know if I should work on covering the new option with tests nonetheless.

**Comment by Gallaecio** [comment]
Could you let me know if I should change something other than fixing conflicts with the `CHANGES` file?

**Comment by Gallaecio** [comment]
@tk0miya Are you the one [I should be bugging](http://www.sphinx-doc.org/en/master/devguide.html#contributing-to-sphinx)? :)

File: doc/usage/extensions/coverage.rst [hunk 1]
```diff
@@ -22,6 +22,14 @@ should check:
 
 .. confval:: coverage_ignore_classes
 
+.. confval:: coverage_ignore_py
```
Comment (tk0miya) [inline doc/usage/extensions/coverage.rst]: `ignore_py` is a bit ambiguous to me. What about `coverage_ignore_pyobject` instead?

File: doc/usage/extensions/coverage.rst [hunk 2]
```diff
@@ -22,6 +22,14 @@ should check:
...
+
+   List of `Python regular expressions`_.
+
+   If any of these regular expressions matches any part of the full import path
+   of a Python API member, that Python API member is excluded from the
```
Comment (tk0miya) [inline doc/usage/extensions/coverage.rst]: "Python API member" is common wording?

File: doc/usage/extensions/coverage.rst [hunk 3]
```diff
@@ -22,6 +22,14 @@ should check:
...
+   documentation coverage report.
+
```
Comment (tk0miya) [inline doc/usage/extensions/coverage.rst:34]: Please add `versionadded:: 2.1` here.

**Review by tk0miya — CHANGES_REQUESTED** [review]
Sorry for late response. LGTM for this idea!

Could you add a testcase to `tests/test_ext_coverage.py` please? New features should be tested.
Thanks,

File: doc/usage/extensions/coverage.rst (hunk 1 above)
Comment (Gallaecio) [reply doc/usage/extensions/coverage.rst]: Switched to `coverage_ignore_pyobjects` (plural to keep it coherent with the other 3 Python options).

File: doc/usage/extensions/coverage.rst (hunk 2 above)
Comment (Gallaecio) [reply doc/usage/extensions/coverage.rst]: Switched to ‘Python objects’. I had no idea what the common wording was :)

File: tests/roots/test-ext-coverage/python.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Undocumented Python objects
+===========================
```
Comment (tk0miya) [inline tests/roots/test-ext-coverage/python.txt]: No results does not mean exclusion working fine. Please add an example not filtered.

Comment (tk0miya) [reply tests/roots/test-ext-coverage/python.txt]: Is there reason to store expected result to file?
I prefer to define it inside test script. How about that?

File: sphinx/ext/coverage.py [hunk 5]
```diff
@@ -130,6 +132,12 @@ def write_c_coverage(self):
                     op.write(' * %-50s [%9s]\n' % (name, typ))
                 op.write('\n')
 
+    def ignore_py(self, full_name):
```
Comment (tk0miya) [inline sphinx/ext/coverage.py]: I think this should be also renamed to `ignore_pyobj`.

**Review by tk0miya — CHANGES_REQUESTED** [review]
LGTM with nits

**Comment by tk0miya** [comment]
In addition, please rebase this to 2.0 branch. master branch is used for 3.0 release now.

File: tests/roots/test-ext-coverage/python.txt (hunk 4 above)
Comment (Gallaecio) [reply tests/roots/test-ext-coverage/python.txt]: Done.

Both the ignored and the non-ignored modules have now the same content, which includes an undocumented class and undocumented methods within a documented class.

**Comment by Gallaecio** [comment]
Rebased.

**Review by tk0miya — APPROVED** [review]
LGTM!

**Comment by tk0miya** [comment]
Merged. Thank you for your contribution!

**Pull request merged**